vllm/worker/multi_step_model_runner.py

```diff
@@ -0,0 +1,506 @@
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union
+
+try:
+    from vllm.attention.backends.flash_attn import FlashAttentionMetadata
+except ModuleNotFoundError:
+    # vllm_flash_attn is not installed, use the identical ROCm FA metadata
+    from vllm.attention.backends.rocm_flash_attn import (
+        ROCmFlashAttentionMetadata as FlashAttentionMetadata)
+
+import torch
+
+from vllm import _custom_ops as ops
+from vllm.distributed import get_pp_group
+from vllm.logger import init_logger
+from vllm.sequence import (CompletionSequenceGroupOutput, IntermediateTensors,
+                           Logprob, SamplerOutput, SequenceGroupMetadata,
+                           SequenceOutput)
+from vllm.worker.model_runner import (GPUModelRunnerBase,
+                                      ModelInputForGPUWithSamplingMetadata)
+from vllm.worker.model_runner_base import (
+    BroadcastableModelInput, _init_attn_metadata_from_tensor_dict,
+    _init_frozen_model_input_from_tensor_dict,
+    _init_sampling_metadata_from_tensor_dict)
+
+from ..model_executor.model_loader.tensorizer import TensorizerConfig
+
+if TYPE_CHECKING:
+    from vllm.attention.backends.abstract import AttentionBackend
+
+logger = init_logger(__name__)
+
+
+@dataclass
+class ModelOutput:
+    """The output of a single model forward pass.
+
+    The sampler_output_ready_event is set when the tensors in
+    sampler_output are ready (the model+sampler forward pass has
+    completed). We use the event to synchronize the GPU->CPU transfer,
+    which we want to only run when the data has been written to the
+    GPU tensors. Until the event is ready, the tensors in sampler_output
+    will have garbage data.
+
+    There are two scenarios:
+    1. The output tensors are ready and we can pythonize them immediately.
+    2. The output tensors are not ready and we need to wait for the event to be
+    ready.
+    """
+    sampler_output: SamplerOutput
+    sampler_output_ready_event: torch.cuda.Event
+    sampled_token_ids: Optional[torch.Tensor] = None
+    pythonized: bool = False
+
+    def pythonize(
```

Comment:
Maybe we could unify the API like this?

```python
def maybe_pythonize(..., blocking: bool):
    if not self.pythonized:
        self.pythonized = self._pythonize_sampler_output(
            input_metadata, copy_stream, pinned_sampled_token_buffer,
            blocking)
        self.pythonized = True

def _pythonize_sampler_output(..., blocking: bool):
        assert self.sampled_token_ids is not None
        if not blocking and not self.sampler_output_ready_event.query():
            return False
        if blocking:
            self.sampler_output_ready_event.synchronize()

        with torch.cuda.stream(copy_stream):
            _pythonize_sampler_output(input_metadata, self.sampler_output,
                                      pinned_sampled_token_buffer,
                                      self.sampled_token_ids)
        return True
```

Further inline `_pythonize_sampler_output` seems also fine but it's up to you.

Reply:
Since these are used within `ModelOutput` anyways, I feel keeping the two behaviors separated is much easier to read and undestand?

Reply:
When I was reviewing these 2 functions, I actually need to diff them using my eyes to know the difference lol
So IMO a single function is actually more clear. Maybe adding some comments if needed.

Reply:
[b8943c1](https://github.com/vllm-project/vllm/pull/7000/commits/b8943c1b6f66a03e3fb03e6dfbc7e7a31648fa71)

Reply:
I kept the `maybe_pythonize` and `pythonize` just so it is clear in the model_runner what is being attempted